src/metabase/upload.clj

```diff
@@ -226,26 +255,11 @@
                           (try (into []
                                      ;; take first two rows and count the number of columns in each to compare headers
                                      ;; vs data rows.
-                                     (comp (take 2) (map count))
+                                     (comp (take max-infered-lines) (map count))
                                      (csv/read-csv reader :separator s))
                                (catch Exception _e nil))))]
     (->> (map (juxt identity count-columns) separators)
-         ;; We cannot have more data columns than header columns
-         ;; We currently support files without any data rows, and these get a free pass.
-         (remove (fn [[_s [header-column-count data-column-count]]]
-                   (when data-column-count
-                     (> data-column-count header-column-count))))
-         ;; Prefer separators according to the follow criteria, in order:
-         ;; - Splitting the header at least once
-         ;; - Giving a consistent column split for the first two lines of the file
-         ;; - The number of fields in the header
-         ;; - The precedence order in how we define them, e.g.. bias towards comma
-         (sort-by (fn [[_ [header-column-count data-column-count]]]
-                    [(when header-column-count
-                       (> header-column-count 1))
-                     (= header-column-count data-column-count)
-                     header-column-count])
-                  u/reverse-compare)
+         (sort-by (comp separator-priority second) u/reverse-compare)
          ffirst
          assert-inferred-separator)))
```

Comment:
Since we're just ranking the separators now, rather than removing them, I think it's safe to remove `assert-inferred-separator` call now